tools/toolchain/gen_mksurf_namelist.py

```diff
@@ -0,0 +1,658 @@
+        for a ctsm case.
+    build_nl:
+        Build the namelist/control file for a ctsm
+        case.
+    """
+
+    def __init__ (self, res, glc_nec, ssp_rcp, crop_flag, input_path, vic_flag, glc_flag, start_year, end_year=None):
+        self.res = res
+        self.glc_nec = glc_nec
+        self.ssp_rcp = ssp_rcp
+        self.crop_flag = crop_flag
+        self.input_path = input_path
+        self.vic_flag = vic_flag
+        self.glc_flag = glc_flag
+        self.start_year = start_year
+        self.end_year = end_year if end_year is not None else start_year
+
+        #-- check if end year value is a valid value
+        self.check_endyear()
+
+        #-- Determine if the case is transient
+        self.check_run_type()
+
+        #-- determine the num_pft
+        self.check_num_pft()
+    
+
+    def __str__(self):
+        return  str(self.__class__) + '\n' + '\n'.join((str(item) + ' = ' + str(self.__dict__[item])
+                    for item in self.__dict__))
+
+    def check_endyear (self):
+        if (int(self.end_year) < int(self.start_year)):
+            sys.exit('ERROR: end_year should be bigger than the start_year : '
+                     + self.start_year.__str__() + '.')
+
+    def check_run_type (self):
+        if (self.end_year > self.start_year):
+            self.run_type = "transient"
+        else:               
+            self.run_type = "timeslice"
+        logging.debug(' run_type  = '+ self.run_type)
+
+    def check_num_pft (self):
+        #-- determine the num_pft
+        if self.crop_flag:
+            self.num_pft      = "78"
+        else:
+            self.num_pft      = "16"
+        logging.debug(' crop_flag = '+ self.crop_flag.__str__()+ ' => num_pft ='+ self.num_pft)
+
+
+    def name_nl  (self):
+        """
+        Build namelist file name.
+        """
+        time_stamp = datetime.today().strftime('%y%m%d')
+        namelist_fname = "surfdata_"+ \
+            self.res+"_"+ \
+            self.ssp_rcp+"_"+ \
+            self.num_pft+ \
+            "pfts_CMIP6_"+ \
+            self.start_year.__str__()+'-' + \
+            self.end_year.__str__()+ \
+            "_c"+time_stamp+".namelist"
+
+        self.namelist_fname  = namelist_fname
+
+    def landuse_filename(self):
+        if (self.run_type == 'transient'):
+            lu_fname = "landuse_timeseries_hist_"+self.num_pft.__str__()+ \
+                       "pfts_simyr"+str(self.start_year)+"-"+str(self.end_year)+".txt"
+        else:
+            lu_fname = ""
+        self.lu_fname = lu_fname
+
+    def create_landuse(self):
+        self.landuse_filename()
+        lu_file = open (self.lu_fname,'w')
+
+        for yr in tqdm.tqdm(range(self.start_year, self.end_year+1)):
+
+            #-- choose different files for years of 850-1850:
+            if (849 < yr < 1850):
+                lu_input_fname = os.path.join(self.input_path, 
+                            "pftcftdynharv.0.25x0.25.LUH2.histsimyr0850-1849.c171012",
```

Comment:
I think it's important to have filenames in data files rather than hardcoded in the code. It could be the XML files as for namelist_defaults_ctsm_tools.xml, or in a different format, but should be in an outside data file. It is good to have it in fewer lines, but we have some development for more flexible data formats that allow this same type of thing in data files. Alper's mechanism for CESM or the new streams mechanism for NUOPC. One of those mechanisms could be used to shorten the number of lines required in the data file.

Reply:
I agree Erik. I think maybe @slevisconsulting can implement this idea.